src/plonk.rs

```diff
@@ -106,6 +120,184 @@ impl<C: Curve> Circuit<C> {
+        let c_plonk_z = pedersen_hash::<C>(&plonk_z_coeffs, &self.msm_precomputation).to_affine();
+
+        // Generate a random alpha from the transcript.
+        let alpha = rescue_hash_n_to_1::<C::BaseField>(vec![beta, c_plonk_z.x, c_plonk_z.y], self.security_bits);
+
+        // Generate the quotient polynomial, t.
+        let plonk_t_points: Vec<C::ScalarField> = todo!();
+        let plonk_t_coeffs = ifft_with_precomputation_power_of_2(&plonk_t_points, &self.fft_precomputation_n);
+        let plonk_t_coeff_chunks: Vec<Vec<C::ScalarField>> = plonk_t_coeffs.chunks(self.degree())
+            .map(|chunk| chunk.to_vec()).collect();
+        let c_plonk_t: Vec<ProjectivePoint<C>> = plonk_t_coeff_chunks.iter()
+            .map(|coeffs| pedersen_hash::<C>(&plonk_t_coeffs, &self.msm_precomputation))
+            .collect();
+        let c_plonk_t = ProjectivePoint::batch_to_affine(&c_plonk_t);
+
+        // Generate a random zeta from the transcript.
+        let zeta_bf = rescue_hash_n_to_1::<C::BaseField>([vec![alpha], flatten_points(&c_plonk_t)].concat(), self.security_bits);
+        let zeta_sf = C::try_convert_b2s(zeta_bf).expect("should fit in both fields with high probability");
+
+        // Open all polynomials at each PublicInputGate index.
+        let num_public_input_gates = ceil_div_usize(self.num_public_inputs, NUM_WIRES);
+        let o_public_inputs = (0..num_public_input_gates)
+            // We place PublicInputGates at indices 0, 2, 4, ...
+            .map(|i| i * 2)
+            .map(|i| self.open_all_polynomials(
+                &wires_coeffs, &plonk_z_coeffs, &plonk_t_coeff_chunks,
+                C::ScalarField::from_canonical_usize(i)))
+            .collect();
+
+        // Open all polynomials at zeta, zeta * g, and zeta * g^65.
+        let o_local = self.open_all_polynomials(&wires_coeffs, &plonk_z_coeffs, &plonk_t_coeff_chunks,
+                                                zeta_sf);
+        let o_right = self.open_all_polynomials(&wires_coeffs, &plonk_z_coeffs, &plonk_t_coeff_chunks,
+                                                zeta_sf * self.subgroup_generator);
+        let o_below = self.open_all_polynomials(&wires_coeffs, &plonk_z_coeffs, &plonk_t_coeff_chunks,
+                                                zeta_sf * self.subgroup_generator.exp_usize(GRID_WIDTH));
+
+        // Get a list of all opened values, to append to the transcript.
+        let all_opening_sets: Vec<OpeningSet<C::ScalarField>> = [
+            o_public_inputs,
+            vec![o_local, o_right, o_below],
+        ].concat();
+        let all_opened_values_sf: Vec<C::ScalarField> = all_opening_sets.into_iter()
+            .map(|os| os.to_vec()).collect::<Vec<_>>()
+            .concat();
+        let all_opened_values_bf: Vec<_> = all_opened_values_sf.into_iter()
+            .map(|f| C::try_convert_s2b(f)
+                .expect("For now, we assume that all opened values fit in both fields"))
```

Comment:
This is a pretty big limitation -- the prover may fail if we try to use -1 as a public input (depending on which field is larger), for example. I'll try to address this soon, but assuming all openings fit in both fields makes things a lot easier for now.